### DISCUSSION TITLE
Freeing mallocs aimed at in stack message causes seg faults

### DISCUSSION THREAD

**Discussion by DaemonDave** [protobuf-c/protobuf-c#649] [Q&A]
I need a variable number of repeated submessages within a protobuf-c message. Everrything works fine until I try to free the malloc sub message contents. If the free() are commented out it exits peacefully. Otherwise, it seg faults.  Is there another way to do this or how do I change this to free elements?

`#include <stdio.h>
# include <stdlib.h>
# include <time.h>
# include <sys/time.h>
# include <sys/timeb.h>
# include <string.h>
# include "config.h"
# include "point32.pb-c.h"
# include "cell.pb-c.h"
# include "header.pb-c.h"
# include <protobuf-c/protobuf-c-text.h>
# include "convert.h"

/*
- message cellinfo
  {
  required uint32 topic_id=1;
  required string topic_name=2;
  required uint32 topic_msg=3;
  required string message_type=4;
  required string md5sum=5;
  required uint32 buffer_size=6;
  required uint32 seq=7; 
  required time stamp=8;
  required string frame_id=9;
  required uint32 cellID=10;
  required point32 centre=11;
  required float population=12;  
  required float density=13;  
  required float area=14;  
  required float neighbours=15;  
  repeated cellsurface surface=16;  

}
- 
-   topic_msg changes how to use the message 
-   ADD_CELL tells client to make/ create and initialize a cell
-   UPDATE_CELL tells client to use data to insert data into cell
-   FUSE_CELL  tells client to conflate data from other sources
-   REQUEST_CELL tells client to send update back to server
  */

/\* 
- This file examimes how to make dynamically allocated messages on heap rather than in 
- the stack to allow variable repeated vars etc.
- 
- I made a simple pseudocode for how one executes this:
- 1 Create message elements using mallocs as needed
- 2 Populating message elements with data
- 3 Pack message
- 4 Free elements
- 5 Free buffer
- 
- 
- */

/*! \def PROTOBUF_C_POINTER_MESSAGE_INIT
-  Initializes a message pointer object once allocated without using PROTOBUF_C_MESSAGE_INIT or custom message init type variant
-  DRE 2016
-  This allows for dynamic list of messages elements for dynamic messages which is the problem I was having for a 
- cellular automata message format that had arbitrary 
-  \* 
- _/
  #define PROTOBUF_C_POINTER_MESSAGE_INIT(msg, msgbase ){ msg->base.descriptor =  &msgbase; msg->base.n_unknown_fields = 0;   msg->base.unknown_fields = NULL;     }
  /_
- ARRAY MESSAGE INIT VERSION
- */
  #define PROTOBUF_C_POINTER_MESSAGE_ARRAY_INIT(msg, i, msgbase ){ msg[i].base.descriptor =  &msgbase; msg[i].base.n_unknown_fields = 0;  msg[i].base.unknown_fields = NULL;   }

/*! \def PROTOBUF_C_POINTER_MALLOC
-  mallocs a pointer of type \* msg
-  */ 
  #ifdef HAVE_MALLOC
  #define PROTOBUF_C_POINTER_MALLOC( msg, type ){ msg = (type *) malloc( sizeof(type) );  }
  #else
  #define PROTOBUF_C_POINTER_MALLOC( msg, type ){ msg = (type *) malloc( sizeof(type) );  }
  #endif
# ifdef HAVE_MALLOC
# define PROTOBUF_C_POINTER_REPEATED_MALLOC( msg, type, num ){ msg = (type *) malloc( sizeof(type) \* num );  }
# else
# define PROTOBUF_C_POINTER_REPEATED_MALLOC( msg, type, num ){ msg = (type *) malloc( sizeof(type) \* num );  }
# endif
# define PROTOBUF_C_POINTER_REPEATED_INIT( msg, msgbase, num ){ PROTOBUF_C_POINTER_MESSAGE_INIT(msg[num], msgbase );  }

/*! \def PROTOBUF_C_POINTER_FREE
-  frees a pointer of \* msg
-  _/ 
  #define PROTOBUF_C_POINTER_FREE_BASE( msg ){ free(msg); }
  /_! \def PROTOBUF_C_POINTER_FREE
-  frees a pointer \* subelem malloc
-  */ 
  #define PROTOBUF_C_POINTER_FREE_SUBELEM( msg,subelem ){ free(msg->subelem); }
  #define PROTOBUF_C_POINTER_FREE_ARRAY_ELEM( msg, elem ){ free(msg[elem]); }
  // there should be a creating spool of code then an unspooling code that frees all the elements for a specific given message
  #define CELLINFO_CREATE_CELLINFO_MESSAGE( msg ){  \
  PROTOBUF_C_POINTER_MALLOC( msg, Cellsurface ) \
  PROTOBUF_C_POINTER_MESSAGE_INIT( msg, cellsurface__descriptor )     \
  PROTOBUF_C_POINTER_MALLOC( msg->surf_centre, Point32    )    \
  PROTOBUF_C_POINTER_MESSAGE_INIT( msg->surf_centre, point32__descriptor )        \
  PROTOBUF_C_POINTER_REPEATED_MALLOC( msg, type, num ) \
   \
  } 
# define CELLINFO_FREE_CELLINFO_MESSAGE( msg ){}

int main (int argc, const char \* argv[]) 
{

  //Cellsurface     cell1 = CELLSURFACE__INIT;      // the test message  
  Cellsurface       \* cell1;        // the test message  

  //Point32                     point = POINT32__INIT;
  // pointer version for messages
  Point32                   \*   point;
  Time                          ts = TIME__INIT;
  Cellinfo          combined= CELLINFO__INIT;       // the supermessage
  struct timeval    stamp;
  int i;
  int k;
  char              text[20];
  void *buf;                     // Buffer to store serialized data
  unsigned len;                  // Length of serialized data

```
char * newstuff;
```

  if (argc <= 3  )
  {   // Allow one or two integers
    fprintf(stderr,"usage: sendmessage - %d  %s  %s  %s\n", argc, argv[1], argv[2], argv[3] );
    return 1;
  }

```
// create message block

// init dynamic point32
//PROTOBUF_C_POINTER_MALLOC( point, Point32 )
//PROTOBUF_C_POINTER_MESSAGE_INIT( point, point32__descriptor ) 
PROTOBUF_C_POINTER_REPEATED_MALLOC( point, Point32*, 11 )   
// make some sample points
```

  for( k = 0; k < 11; k++ )
  {
        PROTOBUF_C_POINTER_MESSAGE_ARRAY_INIT(point, k, point32__descriptor ) 
  } 
  // init dynamic cell surface - must allocate all sub messages in line
  PROTOBUF_C_POINTER_MALLOC( cell1, Cellsurface )

  Cellsurface *\* final; 
  // now making dynamic surfaces per cell
  PROTOBUF_C_POINTER_REPEATED_MALLOC( final, Cellsurface*, 11 )

  // make some sample surfaces
  for( k = 0; k < 11; k++ )
  {
        PROTOBUF_C_POINTER_MALLOC( final[k], Cellsurface    )  
        PROTOBUF_C_POINTER_MESSAGE_INIT(final[k], cellsurface__descriptor ) 
  }

  // populate data into message 

  // repeated submessage xyz coords
  combined.n_surface = 11;  
  combined.surface = final;
  // make some sample surfaces
  for( k = 0; k < 11; k++ )
  {
        combined.surface[k]->surfaceid  = k;
    combined.surface[k]->in         = atoi(argv[2]);  
        combined.surface[k]->out        = atoi(argv[3]);  
  }
  point->x          = atoi(argv[1]);  
  point->y          = atoi(argv[2]);
  point->z          = atoi(argv[3]);
  combined.stamp    = & ts;
  // aim pointer at pre-configured submessage points

  for( k = 0; k < 11; k++ )
  {
    combined.surface[k]->surf_centre =  &point[k]; 
    point[k].x = -1000 +k;
    point[k].y = -2000 +k;
    point[k].z = -3000 +k;  
  }

  // add values into message
  combined.centre       = point;
  combined.population   = 999999;
  combined.density      = 1.54;  
  combined.area         = 85.5444;  
  combined.neighbours   = 4.0; 
  combined.topic_id     = 55;  
  combined.topic_msg    = atoi(argv[1]);
  combined.message_type = atoi(argv[2]);
  combined.topic_name   = atoi(argv[3]);
  combined.md5sum       = atoi(argv[1]);
  combined.cellid       = 20000;

  // you need to initialize the pointer array for the message

  // get time and store it
  gettimeofday(&stamp, NULL);
  ts.secs       = stamp.tv_sec;
  ts.nanosecs   = stamp.tv_usec*1000;

  // LAST STEP SIZE
  len = cellinfo__get_packed_size(&combined);
  // do not mess with buffer_size it is an internal variable set by functions but not
  // documented!

  buf = malloc(len);
  // pack message  
  cellinfo__pack((const Cellinfo *)&combined,buf);

  fprintf(stderr,"Writing %d serialized bytes\n",len); // See the length of message
  newstuff = (char *)protobuf_c_text_to_string((ProtobufCMessage *)&combined, NULL);
  fprintf(stderr,"wrote:[%s]\n", newstuff);  
  // STDOUT to receiver
  fwrite(buf,len,1,stdout); // Write to stdout to allow direct command line piping
  fflush(stdout);
  free(buf); // Free the allocated serialized buffer

  // destroy message mallocs
  PROTOBUF_C_POINTER_FREE_BASE( point )
  PROTOBUF_C_POINTER_FREE_BASE( cell1 )  

  // SEG FAULTS FREEING DYNAMIC MALLOCS ATTACHED TO STACK CELLINFO MESSAGE
  for( k = 0; k < 11; k++ )
  {
        free( final[k] );
  }  
  //Cellsurface *\* final = combined.surface;
  free( final );
  // free some sample surfaces
  //PROTOBUF_C_POINTER_FREE_BASE( combined.surface )  

  return 0;
}
`

**Comment by edmonds** [comment]
https://groups.google.com/d/msg/protobuf-c/TJs1Ui4Z3hg/CYQkHQKyBQAJ
